internal/auth/permissions.go

```diff
@@ -0,0 +1,99 @@
+package auth
+
+import (
+	"net/http"
+	"strings"
+)
+
+const (
+	Forbidden                      = "Forbidden"
+	AuthorizationServerUnavailable = "Authorization server unavailable"
+	BadRequest                     = "Unable to verify request"
+)
+
+func CreatePermissionsMiddleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			// /api, /api/v1 and "auth" don't require permissions
+			parts := strings.Split(r.URL.Path, "/")
+			if len(parts) <= 2 ||
+				parts[2] == "auth" {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// Extract resource and action from the request
+			resource, action := extractResourceAndAction(r)
+			if resource == "" || action == "" {
+				http.Error(w, BadRequest, http.StatusServiceUnavailable)
+				return
+			}
+
+			// TODO [gshilin]: is there a way to handle exceptions automatically?
+			// Maybe we can add x-skip-permission-middleware to openapi.yaml?
+			// Or maybe we don't need this exception and can change the (GET /api/v1/enrollmentconfig) handler instead?
+			if resource == "certificatesigningrequests" && action == "get" {
+				next.ServeHTTP(w, r)
+				return
+			}
+			// Perform permission check
+			allowed, err := GetAuthZ().CheckPermission(r.Context(), resource, action)
+			if err != nil {
+				http.Error(w, AuthorizationServerUnavailable, http.StatusServiceUnavailable)
+				return
+			}
+			if !allowed {
+				http.Error(w, Forbidden, http.StatusForbidden)
+				return
+			}
+
+			// If authorized, proceed to the next handler
+			next.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+func extractResourceAndAction(r *http.Request) (string, string) {
+	// Handle according to the URL structure
+	// e.g., "device", "devices", "devices/{name}", "devices/{name}/sub-action"
+	parts := strings.Split(r.URL.Path, "/")
+	if len(parts) < 3 || len(parts) > 5 {
+		// Empty path
+		// /api
+		// /api/v{version}
+		return "", ""
+	}
+	resource := parts[2]
+	action := ""
+	switch r.Method {
+	case http.MethodGet:
+		action = "get"
+	case http.MethodPost:
+		action = "create"
+	case http.MethodPut:
+		action = "update"
+	case http.MethodPatch:
+		action = "patch"
+	case http.MethodDelete:
+		action = "delete"
+	default:
+		return "", ""
+	}
+	switch len(parts) {
+	case 3: // /api/v{version}/resources
```

Comment:
It seems the counting is not correct.  Since the path starts with "/", then the first part will be an empty string.

Reply:
refurbished parts, now it's correct